File: Data Structures & Algorithms/products-of-array-discluding-self/submission-5.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        pre = [1] * n
        suf = [1] * n
        pre[0] = nums[0]
        suf[n - 1] = nums[n - 1]
        for i in range(1, n - 1):
            pre[i] = pre[i - 1] * nums[i]
            suf[n - 1 - i] = suf[n - i] * nums[n - 1 - i]
        res = []
        for i in range(0, n):
            if i == 0:
                res.append(suf[1])
            elif i == n - 1:
                res.append(pre[n - 2])
            else :
                res.append(suf[i + 1] * pre[i - 1])
        return res

File: Data Structures & Algorithms/palindromic-substrings/submission-5.cpp
class Solution {
public:
    int countSubstrings(string s) {
        int n{s.length()};
        bool dp[n][n]{};
        int res{};
        for (int len{1}; len <= n; ++len) {
            for (int i = 0; i < n - len + 1; ++i) {
                int j{i + len - 1};
                if ((len <= 2 || dp[i + 1][j - 1]) && s[i] == s[j]){
                    res ++;
                    dp[i][j] = true;
                }
            }
        }
        return res;
    }
};
